rfcs/0078-configuration-file-generators.md

```diff
@@ -0,0 +1,146 @@
+approach.
+
+A subtree is established in the Nixpkgs repository serving as a library of
+functions implementing program configuration file generation.
+Each such generator takes as an input a NixOS module system based «subtree»,
+e.g. the values typically bound to `cfg` variable in typical modules.
+
+The output of each such function is an attribute set. The keys are relative
+file names of configuration files, and the values are attribute sets of RFCs#42
+settings abstraction and serialiser.
+
+Along with the function the corresponding type specifications for input and
+output modules are provided.
+
+In some cases we only provide low-level overridable default configuration. In
+this case the input may have the type that is always an empty attribute set,
+`{}`.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+A minimalistic silly example is:
+
+Input:
+```
+  {
+     port = 1234;
+     nodeName = "localhost";
+     content = "Hello";
+  }
+```
+
+Output:
+```
+  {
+    "subdir/listen.conf" = {
+      serialiser = toJSON;
+      settings = {
+        listenPort = "1234";
+        serverGreet = "I am localhost";
+      };
+      type = …;
+    };
+    "subdir/content.conf" = {
+      serialiser = toINI;
+      settings = {
+        greeting = {
+          greeter = "localhost";
+        };
+        payload = {
+          "I say to you: ‘Hello’";
+        };
+      };
+      type = …;
+    };
+  }
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The proposed approach increases the spreading of the related code, from package
+and service module to package and configuration generator and module.
+
+The proposed refactoring incurs a cost in terms of implementation effort across
+NixOS, and likely to create a medium-term situation of partial migration.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Do nothing, continuing with the current code duplication.
+
+Same, but put configuration generators closer to packages. This would mean
+widespread use of the module system inside `pkgs/`. There is also no guarantee
+that all the configuration files describing interaction of multiple software
+packages will have a clear choice of reference package.
+
+There also have been many solutions proposed based on a significant rework of the module system.
+
+Abstract generation of configuration files with package-like flat arguments and
+plain text file outputs. This approach will need less code as long as it we do
+not want type checks, or ease of overriding values.
+
+Implement a complete service abstraction not tied to global system-wide
+assumptions.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Currently none.
+
+# Future work
+[future]: #future-work
+
+Define generic service abstraction for non-`systemd` systems, possibly with a
+convertor from `systemd` units.
+
+Consider some  way of referring from package to related configuration file
```

Comment:
Why? (what for) &rarr; you might want to clarify to improve intelligibility here.

Reply:
Anyone interested in this future work item _now_ is likely to be aware of the context like `passthru.tests`; those who are not familiar with that will wait anyway until some detailed work is started

Reply:
Your response is a little bit enigmatic with regard to rhe matter of the question, but let's make some assumptions at the risk of requiring further clarification:

Do you consider here the possibility of moving service configuration still closer to the packaging itself?

If that is the case, the (new) proposed module system for packaging might come into play. And (very) early coordination with this RFC might benefit future development.

Reply:
We'll see from feedback, but currently the RFC argues against moving configuration generators close to packages. However, we might consider developing a way of having some kind of cross-references available at evaluation time.